pyfr/plugins/turbulence.py

```diff
@@ -0,0 +1,272 @@
+                          (1 - np.cos(ran))*rax @ rax.T)
+
+        self.macro_params = {'ls': ls, 'avgu': avgu, 'yzmin': [ymin, zmin],
+                             'yzdim': [ydim, ydim], 'beta1': beta1,
+                             'beta2': beta2, 'beta3': beta3, 'rot': rot,
+                             'shift': shift, 'ac': ac}
+
+        self.trcl = defaultdict()
+        self.vortbuf = self.getvortbuf()
+        self.vortstructs = self.getvortstructs(intg)
+
+        if not bool(self.vortstructs):
+            self.tnext = float('inf')
+
+    def __call__(self, intg):
+        if intg.tcurr + intg._dt < self.tnext:
+            return
+
+        for etype, [strms, strmsid, buf, tinit, state] in self.vortstructs.items():
+            if self.trcl[etype] > self.tnext:
+                continue
+
+            self.trcl[etype] = self.update_buf(strms, strmsid, buf, self.tnext)
+            tinit.set(buf.tinit)
+            state.set(buf.state)
+
+        self.tnext = min(self.trcl.values())
+
+    def update_buf(self, strms, strmsid, buf, tnext):
+        trcltemp = float('inf')
+        for ele, strm in strms.items():
+            sid = strmsid[ele]
+            if sid >= strm.shape[0]:
+                continue
+
+            tmp = buf[:, ele][buf[:, ele].te > tnext]
+            rem = tmp.shape[0]
+            sft = sid + buf.shape[0] - rem
+            add = rem + strm[sid:sft].shape[0]
+            buf[:rem, ele] = tmp
+            buf[rem:add, ele] = strm[sid:sft]
+            buf[add:, ele] = 0
+
+            if sft < strm.shape[0] and (buf[-1, ele].ts < trcltemp):
+                trcltemp = buf[-1, ele].ts
+
+            strmsid[ele] = sft
+
+        return trcltemp
+
+    def test_nvmx(self, strms, neles, nvmx):
+        tnext = self.tnext
+        strmsid = defaultdict()
+        for ele in strms:
+            strmsid[ele] = 0
+        buf = np.core.records.fromrecords(np.zeros((nvmx, neles)),
+                                          dtype=self.eventdtype)
+
+        while tnext < float('inf'):
+            trcltemp = self.update_buf(strms, strmsid, buf, tnext)
+            if trcltemp < (tnext + self.nvmxtol):
+                return False
+
+            tnext = trcltemp
+
+        return True
+
+    def getvortbuf(self):
+        pcgg = pcg(self.seed)
+
+        vid = 0
+        temp = []
+        tinits = []
+
+        while vid < self.nvorts:
+            tinits.append(self.tstart + 2*self.ls*next(pcgg)[1]/self.avgu)
+            vid += 1
+
+        while any(tinit <= self.tend for tinit in tinits):
+            for vid, tinit in enumerate(tinits):
+                (state, floaty) = next(pcgg)
+                yinit = self.ymin + self.ydim*floaty
+                zinit = self.zmin + self.zdim*next(pcgg)[1]
+                tend = tinit + (2*self.ls/self.avgu)
+                if tend >= self.tbegin and tinit <= self.tend:
+                    temp.append((yinit, zinit, tinit, state))
```

Comment:
Is this the correct state given that you get another result from pcgg at line 172?

Reply:
Yes, I think so, we want the state before it is called twice. We then replay both calls from this state on the device side.